### ISSUE TITLE
Not working on Firebase server

### ISSUE THREAD

**Issue by IdaresitEmmanuel** [redleafsofts/flutter_ffmpeg_wasm#6]
Please help me resolve this issue. It works perfectly during development, but as soon as it is deployed to Firebase server, ffmpeg is unable to even load

firebase.json file

```
{
  "hosting": {
    "public": "build/web",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ],
      "headers": [
        { 
          "source": "**/*.@(html)",
          "headers":[
            {
              "key": "Cross-Origin-Embedder-Policy",
              "value": "require-corp"
            },
            {
          "key": "Cross-Origin-Opener-Policy",
          "value": "same-origin"
            }
        ]
      }
      ]
  }
}
```

index.html contains these headers

```
<script src="https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js" crossorigin="anonymous" async></script>
  <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
  <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
```

**Comment by IdaresitEmmanuel** [comment]
> Please help me resolve this issue. It works perfectly during development, but as soon as it is deployed to Firebase server, ffmpeg is unable to even load
> 
> firebase.json file
> 
> ```
> {
>   "hosting": {
>     "public": "build/web",
>     "ignore": [
>       "firebase.json",
>       "**/.*",
>       "**/node_modules/**"
>     ],
>     "rewrites": [
>       {
>         "source": "**",
>         "destination": "/index.html"
>       }
>     ],
>       "headers": [
>         { 
>           "source": "**/*.@(html)",
>           "headers":[
>             {
>               "key": "Cross-Origin-Embedder-Policy",
>               "value": "require-corp"
>             },
>             {
>           "key": "Cross-Origin-Opener-Policy",
>           "value": "same-origin"
>             }
>         ]
>       }
>       ]
>   }
> }
> ```
> 
> index.html contains these headers
> 
> ```
> <script src="https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js" crossorigin="anonymous" async></script>
>   <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
>   <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
> ```

I works now, but these headers affect every other request in the app and blocks almost all.

**Comment by sarankumar-ns** [comment]
Hi @IdaresitEmmanuel, Can you share your error log that prints on the browser console while loading that page?

If your use case is okay with loading from the static directory then follow the Importing scripts from other domains section at https://github.com/redleafsofts/flutter_ffmpeg_wasm#production.

**Comment by IdaresitEmmanuel** [comment]
@sarankumar-ns There is no error log while loading the page. The problem is that the headers in firebase.json blocked all HTTP requests that are not from the same origin. So I am not able to access files in my Firebase Storage via URL because it is blocked by CORS policy which is needed for flutter_ffmpeg_wasm to work. I had to remove it from the app and use ffmpeg in Cloud Functions to get around it.

**Comment by Megatronicus** [comment]
Did you solve this?

**Comment by Megatronicus** [comment]
I get errors here too 
This seems to be the problem 
<script src="https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js" crossorigin="anonymous" async></script>

I have this in firebase.js, I notice it's different than IdaresitEmmanuel posted a week ago, this sould really be standardized somehow if you could write more precise instructions how to modify this?


{
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "hosting": {
    "public": "build/web",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "headers": [
      {
        "source": "/**",
        "headers": [
          {
            "key": "Cross-Origin-Embedder-Policy",
            "value": "require-corp"
          },
          {
            "key": "Cross-Origin-Opener-Policy",
            "value": "same-origin"
          }
        ]
      }
    ]
  },
  "storage": {
    "rules": "storage.rules"
  },
  "functions": [
    {
      "source": "functions",
      "codebase": "default",
      "ignore": [
        "node_modules",
        ".git",
        "firebase-debug.log",
        "firebase-debug.*.log"
      ],
      "predeploy": [
        "npm --prefix \"$RESOURCE_DIR\" run lint"
      ]
    }
  ]
}


This is what I have in index.html that's inside the web folder of the flutter app. 

In the<head> part 

  <title>TitleOfTheApp</title>
  <link rel="manifest" href="manifest.json">
  <script>
    // The value below is injected by flutter build, do not touch.
    var serviceWorkerVersion = null;
  </script>


  
  <script src="flutter.js" defer></script>
  <script src='https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js'></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/FileSaver.js/1.3.8/FileSaver.min.js">
  </script>

  <script>
        function webSaveAs(blob, name) {
            saveAs(blob, name);
        }
    </script>


  
  <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
  <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">



This is the error I get in the chrome browser console 

GET https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep


Is there a way to maybe include ffmpeg.min.js in the app so it doesn't need to be downloaded every time?

**Comment by Megatronicus** [comment]
Just saw this in your instructions, trying with this now 

The response should contain the proper Cross-Origin-Embedder-Policy: require-corp / Cross-Origin-Opener-Policy: same-origin headers, and the client should add crossorigin="anonymous" to the script element: <script src="..." crossorigin="anonymous" async></script>. Otherwise, you will get the following error:

GET https://unpkg.com/canvaskit-wasm@0.37.1/bin/canvaskit.js net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep 200

In this case, if you don't own the CDN which provides the script, you should download it and serve it from your own CDN that provides those headers or put it in your static directory.



Also in the instructions 

Since Flutter uses external canvaskit source when the app is in production, you should serve your own canvaskit (note: when you are using --web-renderer html this is not a problem):

This is the easiest way to download the canvaskit version which is used by flutter and also defines local canvaskit path:

#!/bin/sh
# Download CanvasKit
flutter build web
canvaskitLocation=$(grep canvaskit-wasm build/web/main.dart.js | sed -e 's|.*https|https|' -e 's|/bin.*|/bin/|' | uniq)
echo "Downloading CanvasKit from $canvaskitLocation"
curl -o build/web/canvaskit.js "$canvaskitLocation/canvaskit.js"
curl -o build/web/canvaskit.wasm "$canvaskitLocation/canvaskit.wasm"
# Configure flutter web to use local canvaskit
flutter build web --dart-define=FLUTTER_WEB_CANVASKIT_URL=/

I'm deploying like this 
flutter build web --web-renderer auto --release

Do I need to use # Download CanvasKit code? If so, do I have to do it once, or every time before a deployment? 

I don't understand this part of the instructions. Could you possibly go into some more detail for this part? 

Note: When importing a script from another domain, it's recommended to use localhost origin or https protocol. Otherwise, you may encounter the following error:

The Cross-Origin-Opener-Policy header has been ignored, because the origin was untrustworthy. It was defined either in the final response or a redirect. Please deliver the response using the HTTPS protocol. You can also use the 'localhost' origin instead. See https://www.w3.org/TR/powerful-features/#potentially-trustworthy-origin and [https://html.spec.whatwg.org/#the-cross-origin-opener-policy-header.](https://html.spec.whatwg.org/#the-cross-origin-opener-policy-header.%22)

Note: When importing a script from another domain, it's recommended to use localhost origin or https protocol. Otherwise, you may encounter the following error:

The Cross-Origin-Opener-Policy header has been ignored, because the origin was untrustworthy. It was defined either in the final response or a redirect. Please deliver the response using the HTTPS protocol. You can also use the 'localhost' origin instead. See https://www.w3.org/TR/powerful-features/#potentially-trustworthy-origin and [https://html.spec.whatwg.org/#the-cross-origin-opener-policy-header.](https://html.spec.whatwg.org/#the-cross-origin-opener-policy-header.%22)

**Comment by Megatronicus** [comment]
I tried 

<script src="..." crossorigin="anonymous" async></script>


appname-g9221.web.app/:61     GET https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js net::ERR_BLOCKED_BY_RESPONSE.NotSameOriginAfterDefaultedToSameOriginByCoep

**Comment by Megatronicus** [comment]
Ok, I've tried putting this into index.html like it is and it appeared on the screen before anything else was loaded, so obviously that's not the way to use that in index.html 


  Cross-Origin-Embedder-Policy: require-corp
  Cross-Origin-Opener-Policy: same-origin

**Comment by Megatronicus** [comment]
I found multiple sites hosting ffmpeg.min.js which could be incorporated into the app instead of made to download in index.html 

https://cdnjs.com/libraries/ffmpeg
https://cdnjs.cloudflare.com/ajax/libs/ffmpeg/0.12.6/umd/ffmpeg.min.js
https://cdn.jsdelivr.net/npm/@ffmpeg/ffmpeg@0.8.3/dist/ffmpeg.min.js

But as you can see, the instructions are different? Which is the right one and how can I use this as a part of the code instead of making it download? For example saving it into the web folder of the flutter app and then referencing that from index.html or from wherever it's needed?

**Comment by Megatronicus** [comment]
Hello? Can you guys please help?

**Comment by Megatronicus** [comment]
I got it to work with the production version in index.html 

```html
<script src="https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js" crossorigin="anonymous" async></script>
``` 

I also used this but then I couldn't load images inside the app anymore, so I'm testing now without that

```html
  <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
  <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
  <script src="..." crossorigin="anonymous" async></script>
```

**Comment by Megatronicus** [comment]
I got it to work using this command, 
flutter build web --web-renderer auto --release
it doesn't have to be 
flutter build web --web-renderer html --release	

In index.html I have this 

```html
  <link rel="manifest" href="manifest.json">
  <script>
    // The value below is injected by flutter build, do not touch.
    var serviceWorkerVersion = null;
  </script>


  <!-- This script adds the flutter initialization JS code -->
  <script src="flutter.js" defer></script>
<!--  <script src='https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js'></script>-->
  <script src="https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js" crossorigin="anonymous" async></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/FileSaver.js/1.3.8/FileSaver.min.js">
  </script>

  <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
  <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
``` 

I found that this was not necessary 
```html
<script src="..." crossorigin="anonymous" async></script>
``` 

However without this part, the images in the app stopped showing

```html
  <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
  <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
``` 

Also, although I did get this error below in the console of brave browser, the trimming and compression still worked 

```dart
BindingError: Expected null or instance of Paragraph, got an instance of Paragraph
main.dart.js:42510     at Error.<anonymous> (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:101:76)
main.dart.js:42510     at new BindingError (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:100:258)
main.dart.js:42510     at P (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:103:249)
main.dart.js:42510     at Xb (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:109:38)
main.dart.js:42510     at cc.bc [as toWireType] (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:112:310)
main.dart.js:42510     at Tb._drawParagraph (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:117:8)
main.dart.js:42510     at a.Canvas.drawParagraph (https://www.gstatic.com/flutter-canvaskit/a794cf2681c6c9fe7b260e0e84de96298dc9c18b/chromium/canvaskit.js:48:336)
main.dart.js:42510     at as0.vP (https://appname-g9221.web.app/main.dart.js:45144:10)
main.dart.js:42510     at aba.b_ (https://appname-g9221.web.app/main.dart.js:90830:9)
main.dart.js:42510     at Q2.b_ (https://appname-g9221.web.app/main.dart.js:94796:5)
``` 


Also aside from all this I do keep getting this error below, about a 10 errors a second in brave browser console only, in chrome it doesn't appear and I don't know if it's even connected with ffmpeg_wasm, I found somewhere it might be connected with .env files which I also have to use Stripe payment and save keys there, need to research this more, just thought I'd write it here to see if you've experienced it in brave browser maybe have a quick fix. 

Another exception was thrown: Instance of 'minified:kS<void>'

**Comment by sarankumar-ns** [comment]
Hi @Megatronicus, Please checkout the configuration used in the example project for reference.

**Comment by Megatronicus** [comment]
> Hi @Megatronicus, Please check out the configuration used in the example project for reference.

Thank you for answering, but you didn't cover a single one of my questions? 

The most important one being 

https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js

How would I save this to a file to be referenced within the code so it doesn't need to be called and downloaded every time in index.html and would that be a good practice?

Please just answer this part.

**Comment by Megatronicus** [comment]
Checked out the official example, you don't even have this inside index.html 

```html
  <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
  <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
```

**Comment by sarankumar-ns** [comment]
> Checked out the official example, you don't even have this inside index.html
> 
> ```
>   <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
>   <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
> ```

Hi @Megatronicus, Header configurations are defined in the `firebase.json` file.

**Comment by Megatronicus** [comment]
> > Checked out the official example, you don't even have this inside index.html
> > ```
> >   <meta http-equiv="Cross-Origin-Embedder-Policy" content="require-corp">
> >   <meta http-equiv="Cross-Origin-Opener-Policy" content="same-origin">
> > ```
> 
> Hi @Megatronicus, Header configurations are defined in the `firebase.json` file.

Thank you for that clarification, but as you can see in the official pub dev https://pub.dev/packages/ffmpeg_wasm
the instructions are 

The document index.html should contain these headers:
```html
Cross-Origin-Embedder-Policy: require-corp
Cross-Origin-Opener-Policy: same-origin
``` 


You can see how that information then would be very confusing. Thank you for making this library but the instructions should really be rewritten and organized better if that is possible. 

Can you please just answer this, I've asked you 2 times already. 

How would I save this to a file to be referenced within the code so it doesn't need to be called and downloaded every time in index.html and would that be a good practice?
https://unpkg.com/@ffmpeg/ffmpeg@0.11.6/dist/ffmpeg.min.js

**Comment by sarankumar-ns** [comment]
<img width="809" alt="Screenshot 2023-10-07 at 10 21 44 AM" src="https://github.com/redleafsofts/flutter_ffmpeg_wasm/assets/33630797/1fdf0d0e-8410-4d86-9e5f-747f7aef3922">

This issue #6 is related to a hosting error on Firebase Hosting. The documentation clearly says that for the Firebase Hosting, you need to set up headers in `firebase.json`. That's where you configure headers for the Firebase hosting server. So you don't need to add that header in `index.html`.


And regarding this,

> How would I save this to a file to be referenced within the code so it doesn't need to be called and downloaded every time in index.html and would that be a good practice?

 Please open a new issue for this. So, it will be easy to follow through with the issue.

**Comment by Megatronicus** [comment]
Just to reiterate because this is really confusing me, where does this part go in the index.html  file, in the head or body and after or before which parts in the head or the body? I tried putting it in the body part of index.html but then it showed those words before the app started on the blank screen and only those words, so obviously that's not the way to go. 

/
```html
Cross-Origin-Embedder-Policy: require-corp
Cross-Origin-Opener-Policy: same-origin
``` 
into index.html like that, it doesn't work, where should it go if it goes like that, in the head part?

**Comment by sarankumar-ns** [comment]
Hi @Megatronicus,

To know more about headers Please refer to these:

- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta#http-equiv
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers
- https://stackoverflow.com/questions/36159391/can-we-add-http-response-headers-directly-into-html-pages

If you're using Firebase then you don't need to set these headers in `index.html`. Instead, you need to configure it in firebase.json.
